edwards/ed25519/eddsa.go

```diff
@@ -0,0 +1,154 @@
+package ed25519
+
+import (
+	"crypto/cipher"
+	"crypto/sha512"
+	"errors"
+	"fmt"
+
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/nist"
+	"github.com/dedis/crypto/random"
+	"github.com/dedis/crypto/util"
+)
+
+type EdDSA struct {
+	Secret abstract.Secret
+	prefix []byte
+	Public abstract.Point
+}
+
+func NewEdDSA(randStream cipher.Stream) *EdDSA {
+	// XXX if this PR is accepted, then change this line to iterate as long as
+	// we have 0x00...00 slice buffer. See issue #70
+	randSlice := random.Bytes(32, randStream)
+	hash := sha512.New()
+	hash.Write(randSlice)
+	// XXX Should we copy it to s and prefix ?
+	expanded := hash.Sum(nil)
+	s := expanded[0:32]
+	prefix := expanded[32:]
+	s[0] &= 0xf8
+	s[31] &= 0x7f
+	s[31] |= 0x40
+	// giving the reverse as nist.Int expect big endian
+	var reverse [32]byte
+	util.Reverse(reverse[:], s)
+	suite := NewAES128SHA256Ed25519(false)
+	// We can't use Secret().Pick() because it uses random.int which tests if
+	// the results is < Mod, which is never true (because of the bit tweaks)
+	sec := suite.Secret().(*nist.Int)
+	sec.SetLittleEndian(s)
+	pub := suite.Point().Mul(nil, sec)
+	return &EdDSA{
+		Secret: sec,
+		Public: pub,
+		prefix: prefix,
+	}
+}
+
+func (ed *EdDSA) Sign(msg []byte) ([]byte, error) {
+	sec := ed.Secret
+	hash := sha512.New()
+	hash.Write(ed.prefix)
+	hash.Write(msg)
+
+	suite := NewAES128SHA256Ed25519(false)
+
+	// deterministic random secret and its commit
+	r := suite.Secret().(*nist.Int)
```

Comment:
Well we would still have to put them in big endian before, so either 
```
buffBE := util.Reverse(buffBE,buff)
secret.SetBytes(buffBE)
```
or
```
sNist := secret.(*nist.Int)
sNist.SetLittleEndian(buffer)
```
Same amount of code and for me it would be confusing to have `UnmarshalBinary(b []byte)` and at the same time `SetBytes(b []byte)`. There's no clear separate logic between the two...

For me one useful addition would be to specify the endianess on a nist.Int and having `Marshal/UnmarshalBinary` return the correct endianess.